toolchain/check/import_ref.cpp

```diff
@@ -22,29 +22,30 @@ namespace Carbon::Check {
 //
-// 1. If Work has received a constant, it's considered resolved.
-//    - If made_forward_decl, resolve unconditionally.
+// 1. If a constant value is already known for the work item, and we're
+//    processing it for the first time, it's considered resolved.
 //    - The constant check avoids performance costs of deduplication on add.
+//    - If `retry` is set, we process it again, because it didn't complete last
+//      time, even though we have a constant value already.
 // 2. Resolve the instruction: (TryResolveInst/TryResolveTypedInst)
-//    - For most cases:
-//      A. For types which _can_ be forward declared, when not
-//         made_forward_decl:
-//        i. Start by making a forward declared type to address circular
-//           references.
-//        ii. If the imported type is not defined, return the constant.
-//        iii. Otherwise, set made_forward_decl and continue resolving.
-//          - Creating a forward declaration of the type will have set the
-//            constant, which influences step (1); setting made_forward_decl
-//            gets us a second resolve pass when needed.
-//      B. Gather all input constants.
-//        - Gathering constants directly adds unresolved values to work_stack_.
-//      C. If any need to be resolved (HasNewWork), return Invalid; this
-//         instruction needs two calls to complete.
-//      D. Build any necessary IR structures, and return the output constant.
-//    - For trivial cases with zero or one input constants, this may return
-//      a constant (if one, potentially Invalid) directly.
-// 3. If resolving returned a non-Invalid constant, pop the work; otherwise, it
-//    needs to remain (and may no longer be at the top of the stack).
+//    - For instructions that can be forward declared, if we don't already have
+//      a constant value from a previous attempt at resolution, start by making
+//      a forward declared constant value to address circular references.
+//    - Gather all input constants.
+//      - Gathering constants directly adds unresolved values to work_stack_.
+//    - If any need to be resolved (HasNewWork), return Retry(): this
+//      instruction needs two calls to complete.
+//      - If the constant value is already known because we have made a forward
+//        declaration, pass it to Retry(). It will be passed to future attempts
+//        to resolve this instruction so the earlier work can be found, and will
+//        be made available for other instructions to use.
+//      - The second attempt to resolve this instruction must produce the same
+//        constant, because the value may have already been used by resolved
+//        instructions.
+//    - Build any necessary IR structures, and return the output constant.
+// 3. If resolve didn't return Retry(), pop the work. Otherwise, it needs to
+//    remain, and may no longer be at the top of the stack; set `retry` on it so
+//    we'll make sure to run it again later.
```

Comment:
@jonmeow I had thought these changes were necessary for correctness, but now that I've finished debugging the issues I was seeing, I don't think they are. However, I do prefer where I ended up: returning a retry / done flag seems nicer to me than carrying that information partly in whether `Invalid` is returned and partly in whether `made_forward_decl` is set.

I'm happy to roll back the refactoring here if you don't like it so much, or split it out to a separate PR. Let me know what  you'd prefer.

Reply:
I still see code like `work_stack_.back().retry = true;`. This makes it look like information is still only partly in the return value, but there's more code now to understand. Can that be improved, so that the only interaction with work_stack_ is here?